datasets/stream_spec.py

```diff
@@ -0,0 +1,192 @@
+                selected_classes.append(label)
+
+            selected_classes.remove(None)
+
+            for c in selected_classes:
+                classes_list.remove(c)
+            task_classes.append(list(selected_classes))
+
+        # assign remaining classes to the first task
+        for c in classes_list:
+            if c not in task_classes[0]:
+                task_classes[0].append(c)
+
+        class_mapping = dict()
+        last_class = 0
+        for task_class in task_classes:
+            for c in task_class:
+                if c not in class_mapping:
+                    class_mapping[c] = last_class
+                    last_class += 1
+
+        new_task_classes = list()
+        for task_class in task_classes:
+            new_classes = sorted(class_mapping[c] for c in task_class)
+            new_task_classes.append(new_classes)
+        return new_task_classes
+
+    def __next__(self):
+        while self.current_task < self.n_tasks:
+            yield self.task_classes[self.current_task]
+            self.current_task += 1
+
+    def __iter__(self):
+        return next(self)
+
+    @property
+    def drifted_classes(self):
+        """return all classes for which drift has occured so far"""
+        seen_classes = set()
+        drifted_cl = set()
+        for i in range(min(self.current_task+1, self.n_tasks)):
+            for c in self.task_classes[i]:
+                if c in seen_classes:
+                    drifted_cl.add(c)
+                else:
+                    seen_classes.add(c)
+        return list(drifted_cl)
+
+    @property
+    def drifted_classes_last_task(self):
+        """return classes from last task for which dirft has occured"""
+        current_classes = self.task_classes[self.current_task]
+        current_classes = set(current_classes)
+        drifted_classes = set(self.drifted_classes)
+        current_drifted = set.intersection(current_classes, drifted_classes)
+        return list(current_drifted)
+
+    @property
+    def new_classes_last_task(self):
+        """return classes from current task"""
+        current_classes = self.task_classes[self.current_task]
+        current_classes = set(current_classes)
+        drifted_classes = set(self.drifted_classes)
+        current_drifted = current_classes - drifted_classes
+        return list(current_drifted)
+
+    @property
+    def max_drifts_per_class(self):
+        counter = collections.Counter()
+        for classes in self.task_classes:
+            counter.update(classes)
+        max_occurence = counter.most_common(1)[0][1]
+        return max_occurence - 1
+
+
+if __name__ == '__main__':
+    s = StreamSpecification(5, 2, 10, 45, n_drifts=1)
+    for task_classes in s:
+        print('task_classes = ', task_classes)
+        print('drifted_classes = ', s.drifted_classes)
+        print('drifted_classes_last_task = ', s.drifted_classes_last_task)
+        print()
```

Comment:
Was this part intended for testing?

Reply:
Yes. In general if I left somewhere if __name__ == '__main__' statement it was for testing. Here code inside if was moved into separate test file, so can be removed.